src/engine/Union.cpp

```diff
@@ -43,9 +43,16 @@ Union::Union(QueryExecutionContext* qec,
       _columnOrigins[it.second.columnIndex_][1] = NO_COLUMN;
     }
   }
-  AD_CORRECTNESS_CHECK(ql::ranges::all_of(_columnOrigins, [](const auto& el) {
-    return el[0] != NO_COLUMN || el[1] != NO_COLUMN;
-  }));
+  // Make sure that the column origins are valid. Because later down the line we
+  // might perform unchecked access using these indices.
+  AD_CORRECTNESS_CHECK(
+      ql::ranges::all_of(_columnOrigins, [this](const auto& el) {
+        return (el[0] != NO_COLUMN || el[1] != NO_COLUMN) &&
+               (el[0] == Union::NO_COLUMN ||
+                el[0] < _subtrees[0]->getResultWidth()) &&
+               (el[1] == Union::NO_COLUMN ||
+                el[1] < _subtrees[1]->getResultWidth());
```

Comment:
hard to read and maintain.
introduce small helper lambdas, s.t. you here write
`return atLeastOneDefiend(el) && isValid(el[0], subtrees[0]) && isValid(el[1], subtrees[1])`.
Or even better, let the `isValid` thing here be a loop (el.size() must be same as subtrees.size, and then use `::"ranges::zip` or something.

That makes this place stable even even we implement nary unions at some point.